src/linker/pipeline.py

```diff
@@ -87,3 +45,90 @@ def _validate_implementations(self) -> Dict:
             if implementation_errors:
                 errors["IMPLEMENTATION ERRORS"][implementation.name] = implementation_errors
         return errors
+
+    @property
+    def implementation_indices(self) -> dict:
+        return {
+            implementation.name: idx
+            for idx, implementation in enumerate(self.implementations)
+        }
+
+    # The following are in Pipeline instead of Implementation because they require
+    # information about the entire pipeline (namely the index number),
+    # not just the individual implementation.
+
+    def get_step_id(self, implementation: Implementation) -> str:
+        idx = self.implementation_indices[implementation.name]
+        step_number = str(idx + 1).zfill(len(str(len(self.implementations))))
+        return f"{step_number}_{implementation.step_name}"
+
+    def get_input_files(self, implementation: Implementation, results_dir: Path) -> list:
```

Comment:
it would probably make sense to put `results_dir` in the config, so we don't have to pass this basically constant path around all over the place

Reply:
I agree. I think it ended up this way b/c I needed the results dir to be known from cli `linker run-slurm-job` and I didn't want to deal w/ passing Config around back to command line. But maybe it's worth sorting out.

Reply:
I'd prefer to do this in a separate refactor--I'll file a ticket for it.